# **WEEK 5 Assignment – Introduction to Deep Learning & Neural Networks with Keras**
Surya Prakash Karasala
 
Date: 03/22/2021

## **Importing Required Packages**

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

## Loading Data

Let's read the dataset into a pandas dataframe.

In [2]:
concrete_data = pd.read_csv('https://cocl.us/concrete_data')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

###**Let's check how many data points we have**.

In [3]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [4]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [5]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

### **Cleaning and Normalizing Data**

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [6]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [7]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [8]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [9]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [11]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [12]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [13]:
from keras.models import Sequential
from keras.layers import Dense

In [14]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [15]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [16]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [17]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [18]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 1594.0135
Epoch 2/50
23/23 - 0s - loss: 1576.2075
Epoch 3/50
23/23 - 0s - loss: 1558.5845
Epoch 4/50
23/23 - 0s - loss: 1541.2576
Epoch 5/50
23/23 - 0s - loss: 1523.7705
Epoch 6/50
23/23 - 0s - loss: 1506.1036
Epoch 7/50
23/23 - 0s - loss: 1488.3081
Epoch 8/50
23/23 - 0s - loss: 1469.9579
Epoch 9/50
23/23 - 0s - loss: 1450.9972
Epoch 10/50
23/23 - 0s - loss: 1431.7112
Epoch 11/50
23/23 - 0s - loss: 1411.5498
Epoch 12/50
23/23 - 0s - loss: 1390.8147
Epoch 13/50
23/23 - 0s - loss: 1369.4944
Epoch 14/50
23/23 - 0s - loss: 1347.3246
Epoch 15/50
23/23 - 0s - loss: 1324.6793
Epoch 16/50
23/23 - 0s - loss: 1300.9048
Epoch 17/50
23/23 - 0s - loss: 1276.8055
Epoch 18/50
23/23 - 0s - loss: 1252.0074
Epoch 19/50
23/23 - 0s - loss: 1225.9170
Epoch 20/50
23/23 - 0s - loss: 1199.3163
Epoch 21/50
23/23 - 0s - loss: 1171.7528
Epoch 22/50
23/23 - 0s - loss: 1143.6730
Epoch 23/50
23/23 - 0s - loss: 1115.5358
Epoch 24/50
23/23 - 0s - loss: 1086.2753
Epoch 25/50
23/23 - 0s - 

Next we need to evaluate the model on the test data.

In [19]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 350.0945


350.0944519042969

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [20]:
from sklearn.metrics import mean_squared_error

In [21]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

350.09445260566974 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [22]:
total_mean_squared_errors = 50
epochs = 100
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 108.60173797607422
MSE 2: 80.59101104736328
MSE 3: 45.48329162597656
MSE 4: 41.37382888793945
MSE 5: 41.18642044067383
MSE 6: 41.631282806396484
MSE 7: 43.18836975097656
MSE 8: 32.86707305908203
MSE 9: 35.026668548583984
MSE 10: 35.419010162353516
MSE 11: 36.40480041503906
MSE 12: 32.33447265625
MSE 13: 38.09294509887695
MSE 14: 42.285430908203125
MSE 15: 32.784725189208984
MSE 16: 27.979328155517578
MSE 17: 33.61988067626953
MSE 18: 32.86384201049805
MSE 19: 33.38349151611328
MSE 20: 34.4528923034668
MSE 21: 32.2415885925293
MSE 22: 30.170028686523438
MSE 23: 29.17970085144043
MSE 24: 30.176422119140625
MSE 25: 31.18229866027832
MSE 26: 34.4509162902832
MSE 27: 28.73182487487793
MSE 28: 32.17691421508789
MSE 29: 34.103965759277344
MSE 30: 32.48359298706055
MSE 31: 26.585941314697266
MSE 32: 27.853954315185547
MSE 33: 27.680749893188477
MSE 34: 30.393199920654297
MSE 35: 33.23678207397461
MSE 36: 36.36674880981445
MSE 37: 25.92095947265625
MSE 38: 30.62775993347168
MSE 39: 30.11